src/cargo/util/toml.rs

```diff
@@ -428,11 +428,20 @@ impl TomlManifest {
 
         let blacklist = vec!["build", "deps", "examples", "native"];
 
+        let mut bin_names: Vec<String> = vec![];
+
         for bin in bins.iter() {
             if blacklist.iter().find(|&x| *x == bin.name()) != None {
                 bail!("the binary target name `{}` is forbidden",
                       bin.name())
             }
+
+            if bin_names.iter().find(|&x| *x == bin.name()) != None {
+                bail!("rejected binary name {}: only unique binary names are accepted",
```

Comment:
Perhaps something along the lines of:

> found duplicate binary name `{}`, but all binary targets must have a unique name

Could you also extend this to cover examples as well?